pkg/util/ip_address.go

```diff
@@ -24,28 +28,37 @@ func ParseIPAddress(input string) string {
 
 // SplitHostPortDefault splits ip address/hostname string by host and port. Defaults used if no match found
 func SplitHostPortDefault(input, defaultHost, defaultPort string) (host string, port string) {
-	port = defaultPort
-	s := input
-	lastIndex := strings.LastIndex(input, ":")
-
-	if lastIndex != -1 {
-		if lastIndex > 0 && input[lastIndex-1:lastIndex] != ":" {
-			s = input[:lastIndex]
-			port = input[lastIndex+1:]
-		} else if lastIndex == 0 {
-			s = defaultHost
-			port = input[lastIndex+1:]
+	start := 0
+	// Determine if IPv6 address, in which case IP address will be enclosed in square brackets
+	if strings.Index(input, "[") == 0 {
+		addrEnd := strings.LastIndex(input, "]")
+		if addrEnd < 0 {
+			// Malformed address
+			return defaultHost, defaultPort
 		}
-	} else {
-		port = defaultPort
+
+		start = addrEnd
+	}
+	if strings.LastIndex(input[start:], ":") < 0 {
+		// There's no port section of the input
+		// It's still useful to call net.SplitHostPort though, since it removes IPv6
+		// square brackets from the address
+		input = fmt.Sprintf("%s:%s", input, defaultPort)
 	}
 
-	s = strings.Replace(s, "[", "", -1)
-	s = strings.Replace(s, "]", "", -1)
-	port = strings.Replace(port, "[", "", -1)
-	port = strings.Replace(port, "]", "", -1)
+	host, port, err := net.SplitHostPort(input)
+	if err != nil {
+		panic(fmt.Sprintf("net.SplitHostPort failed for '%s': %v", input, err))
```

Comment:
Why a call to panic?

Reply:
@kylebrandt Is there any sensible way for us to recover from this error? Ideally I would change `SplitHostPortDefault` to return an error, so we can propagate it.

Reply:
I think the implication in `SplitHostPortDefault` is that the defaults will be used in the case of a parse error - "no match".

Reply:
We might want @marefr to weigh in

Reply:
> I think the implication in `SplitHostPortDefault` is that the defaults will be used in the case of a parse error - "no match".

OK, not sure what the original intention is.

Reply:
> I think the implication in SplitHostPortDefault is that the defaults will be used in the case of a parse error - "no match".

Yes, that's the idea. For the `UserAuthTokenService` we don't want to fail a user to login/rotate auth token because the ip address couldn't be determined. But introducing return of errors is maybe a way forward and let the caller decide what to do?

Reply:
@marefr I think returning an error and letting caller decide does sound better 👍